File: app/src/main/java/com/falcon/restaurants/screens/splash/SplashViewModel.java
package com.falcon.restaurants.screens.splash;

import android.annotation.SuppressLint;
import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import com.falcon.restaurants.screens.restaurant.RestaurantViewModel;
import com.falcon.restaurants.screens.meal.MealViewModel;
import com.falcon.restaurants.utils.Logger;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.BiFunction;
import io.reactivex.schedulers.Schedulers;

public class SplashViewModel extends AndroidViewModel {

    private static final String TAG = "SplashViewModel";

    RestaurantViewModel restaurantViewModel;
    MealViewModel mealViewModel;

    public SplashViewModel(@NonNull Application application,
                           RestaurantViewModel restaurantViewModel,
                           MealViewModel mealViewModel
                           ) {
        super(application);
        this.restaurantViewModel = restaurantViewModel;
        this.mealViewModel = mealViewModel;
    }

    public interface AllUpsertedListener{
        void onSuccess();
        void onFailed(Throwable e);
    }

    @SuppressLint("CheckResult")
    public void fetchAllData(AllUpsertedListener allUpsertedListener){

       Observer<String> observer = new Observer<String>() {
           @Override
           public void onSubscribe(@io.reactivex.annotations.NonNull Disposable d) {
               Logger.log( TAG, "onSubscribe: ");
           }

           @Override
           public void onNext(@io.reactivex.annotations.NonNull String string) {
               Logger.log( TAG, "onNext: " + string);
           }

           @Override
           public void onError(@io.reactivex.annotations.NonNull Throwable e) {
               Logger.log( TAG, "onError: " + e.getLocalizedMessage());
               allUpsertedListener.onFailed(e);
           }

           @Override
           public void onComplete() {
               Logger.log(TAG,  "onComplete: ");
               allUpsertedListener.onSuccess();
           }
       };

        Logger.log( TAG, "startDownload: ");

        Observable.zip(
                mealViewModel.fetch(),
                restaurantViewModel.fetch(),
                new BiFunction<String, String, String>() {
                    @io.reactivex.annotations.NonNull
                    @Override
                    public String apply(@io.reactivex.annotations.NonNull String s,
                                        @io.reactivex.annotations.NonNull String s2) throws Exception {
                        Logger.log( TAG, "apply: s: " + s + " , s2: " + s2);
                        return s + " " + s2;
                    }
                }).subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io()).subscribe(observer);
    }

}

File: app/src/main/java/com/falcon/restaurants/screens/splash/SplashViewModel.kt
package com.falcon.restaurants.screens.splash

import android.annotation.SuppressLint
import android.app.Application
import androidx.annotation.NonNull
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import com.falcon.restaurants.screens.restaurant.RestaurantViewModel
import com.falcon.restaurants.screens.meal.MealViewModel
import com.falcon.restaurants.utils.Logger

import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import io.reactivex.schedulers.Schedulers

class SplashViewModel (val application: Application,
                       val restaurantViewModel: RestaurantViewModel,
                       val mealViewModel: MealViewModel
) : ViewModel() {

    val TAG: String = "SplashViewModel"


    interface AllUpsertedListener{
        fun onSuccess()
        fun onFailed(e: Throwable)
    }

    @SuppressLint("CheckResult")
    fun fetchAllData(allUpsertedListener: AllUpsertedListener){

        val observer: Observer<String> = object : Observer<String> {
            override fun onSubscribe(d: Disposable) {
                Logger.log( TAG, "onSubscribe: ")
            }

            override fun onNext(string: String) {
                Logger.log( TAG, "onNext: " + string)
            }

            override fun onError(e: Throwable) {
                Logger.log( TAG, "onError: " + e.getLocalizedMessage())
                allUpsertedListener.onFailed(e)
            }

            override fun onComplete() {
                Logger.log(TAG,  "onComplete: ")
                allUpsertedListener.onSuccess()
            }

        }

        Logger.log( TAG, "startDownload: ")

        Observable.zip(
                mealViewModel.fetch(),
                restaurantViewModel.fetch(),
                object : BiFunction<String, String, String>{
                    @io.reactivex.annotations.NonNull
                    override fun apply(@io.reactivex.annotations.NonNull s: String,
                                        @io.reactivex.annotations.NonNull s2: String): String {
                        Logger.log( TAG, "apply: s: " + s + " , s2: " + s2)
                        return s + " " + s2
                    }
                }).subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io()).subscribe(observer)
    }

}
